### PR TITLE
feat: Redirect to requested URL after SSO login

### PR DESCRIPTION
**Pull request by simster7**
Depends on: https://github.com/argoproj/argo/pull/4496

Closes: https://github.com/argoproj/argo/issues/4370

Signed-off-by: Simon Behar <simbeh7@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: server/auth/sso/sso.go [hunk 1]
```diff
@@ -180,10 +180,11 @@ func newSso(
 const stateCookieName = "oauthState"
 
 func (s *sso) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	redirectUrl := r.URL.Query().Get("redirect")
 	state := pkgrand.RandString(10)
 	http.SetCookie(w, &http.Cookie{
-		Name:     stateCookieName,
-		Value:    state,
+		Name:     state,
```
Comment (simster7) [inline server/auth/sso/sso.go:186]: https://auth0.com/docs/protocols/state-parameters#use-the-stored-url-to-redirect-users recommends using the nonce/state as the name of the cookie for added security

File: server/auth/sso/sso.go [hunk 2]
```diff
@@ -195,18 +196,13 @@ func (s *sso) HandleRedirect(w http.ResponseWriter, r *http.Request) {
...
-	cookie, err := r.Cookie(stateCookieName)
-	http.SetCookie(w, &http.Cookie{Name: stateCookieName, MaxAge: 0})
+	cookie, err := r.Cookie(state)
+	http.SetCookie(w, &http.Cookie{Name: state, MaxAge: 0})
 	if err != nil {
 		w.WriteHeader(400)
 		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %v", err)))
 		return
 	}
-	if state != cookie.Value {
```
Comment (simster7) [inline server/auth/sso/sso.go:205]: No need to check this anymore, since the name of the cookie is the state

**Comment by simster7** [comment]
Test failure is unrelated: https://github.com/argoproj/argo/issues/4501

**Review by alexec — APPROVED** [review]
LGTM. Maybe make an engineering build and ask community member to try it out?

**Comment by simster7** [comment]
@alexec I like that suggestion, will do.

**Pull request merged**